### ISSUE TITLE
Run a nightly build on GCP to catch failures

### ISSUE THREAD

**Issue by stephenplusplus** [googleapis/google-cloud-common#222]
Originally proposed by @lukesneeringer [here](https://github.com/GoogleCloudPlatform/google-cloud-node/issues/1860#issuecomment-288550644) (feel free to add more detail):

> If it is critical [to test on OS X], then I would probably ask that we shift our focus to using CI for unit tests and having GCE instances that run unit and system tests on each type of platform.

**Comment by stephenplusplus** [comment]
Also, if this issue has been discussed before (it's starting to feel like we've discussed *everything* before), please link and close this one!

**Comment by lukesneeringer** [comment]
So, this is a huge can of worms, but:

  * I certainly would not ditch CircleCI. Running the unit tests on a single platform on pull requests seems critical to me. Running system tests on pushes to master is also critical. (I would like to run relevant system tests on PRs, but secret handling makes that not safe.)

  * It might be reasonable to shift to a nightly build where we build custom Windows/OSX/Linux GCE boxes -- spin one up, run the tests, send an email on a failure, and spin them back down again. What this would basically look like would be having a master GCE instance with something like Ansible installed, and an Ansible playbook to create the instances, run the appropriate thing on the instances, and then delete them on completion.

This is a non-trivial task and I am not advocating doing it "right now".

**Comment by stephenplusplus** [comment]
> (I would like to run relevant system tests on PRs, but secret handling makes that not safe.)

I had a thought on that the other day-- I haven't looked into this, but if someday GH were to open a hook for a "Reviewers Approved" status (or maybe a hook triggered by us adding an "Approved" label), it could trigger a build that would run the system tests. That would make a whole lot more sense than "Make sure system tests pass *after* you merge the code".

> It might be reasonable to shift to a nightly build where we build custom Windows/OSX/Linux GCE boxes -- spin one up, run the tests, send an email on a failure, and spin them back down again.

That would be awesome, and would resolve #199. Which I'll close now, since this issue has more meat.

**Comment by lukesneeringer** [comment]
> I had a thought on that the other day-- I haven't looked into this, but if someday GH were to open a hook for a "Reviewers Approved" status (or maybe a hook triggered by us adding an "Approved" label), it could trigger a build that would run the system tests. That would make a whole lot more sense than "Make sure system tests pass after you merge the code".

Yeah, I would _really_ like a way -- and this is really a feature _CircleCI_ needs, not a feature GitHub needs -- but I would really like a way to say, "these four forks are special whitelisted forks, and you can pass secrets to them".

But, that does not exist yet.

**Comment by dhermes** [comment]
@jonparrott's DPE-bot may be able to do that

**Comment by lukesneeringer** [comment]
> Yeah, I would really like a way -- and this is really a feature CircleCI needs, not a feature GitHub needs -- but I would really like a way to say, "these four forks are special whitelisted forks, and you can pass secrets to them".

> But, that does not exist yet.

That said, we could totally get around that limitation by using branches pushed to `GoogleCloudPlatform/repo-name` rather than `our-github-user/repo-name`. That also scales better because anyone with write bit can do it.

**Comment by stephenplusplus** [comment]
Can you elaborate on that more? How would the process look, starting from:

- @&shy;supercoolcontributor sends us a PR
- Circle CI / AppVeyor runs unit tests, they pass
- ??

**Comment by jonparrott** [comment]
re: nightly builds on Windows / Linux

DevRel has a Jenkins cluster we use for nightly and large system tests that we're willing to share with the API team. There is one big caveat - it is *heavily* firewalled because of security concerns. You can only access it from Google IPs (you must be on campus or on VPN).

Also we don't currently have OS X workers, but it might be possible to look into that.

/cc @waprin

**Comment by lukesneeringer** [comment]
So, for our internal work, basically we would start pushing to branches on the main repo instead of our own. So, right now, if I want to make a change, I do this...

```bash
$ git checkout -b changeset-branch
$ [.. make changes ...]
$ git commit -a -m "My cool changes."  # Yes, I know, commit -a is discouraged. It is an example. :-)
$ git push -u lukesneeringer changeset-branch
```

And `lukesneeringer` is an origin set to `git@github.com:lukesneeringer/google-cloud-node.git`.

Instead of that last line, I would do `git push -u public changeset-branch`, where `public` is an origin set to `git@github.com:GoogleCloudPlatform/google-cloud-node.git`.

Now, when I make the PR, it is still part of the same repo, and the system tests run.

If someone who _is not_ part of our team (and therefore does not have commit bit on the repo) does a PR, they would do it with the process we currently use. The system tests just would not run until after we committed it (unless we checked out the branch and ran it manually).

Basically what this would mean is: at least the system tests run for us. They still would not run for outside contributors. Still a step up from what we have now, which is that they run for nobody.

**Comment by lukesneeringer** [comment]
@jonparrott Interesting. Good to know -- thanks.

**Comment by daspecster** [comment]
CircleCI has OSX, but it looks like limited versions of python installed on it...however I imagine virtualenvs might solve that?

https://circleci.com/mobile/osx/ <-- it says mobile but...

>Test your app on our dedicated OS X cloud
Running a dedicated build box is expensive and time consuming. Let us manage the build environment with our dedicated Apple hardware infrastrastructure. You focus on building great OS X apps.

**Comment by blowmage** [comment]
The Ruby repo has OS X tests. It uses CircleCI as the main build, because it is so much faster than Travis. We have Travis set up to build master only, OS X only. And AppVeyor is similarly set up to build master only on Windows.

**Comment by lukesneeringer** [comment]
> CircleCI has OSX, but it looks like limited versions of python installed on it...however I imagine virtualenvs might solve that?

CircleCI 2.0 has moved entirely to Docker anyway, so it can have whatever versions of Python (or whatever else) that the Docker image has.

**Comment by daspecster** [comment]
@lukesneeringer their OSX builds are all docker?

**Comment by lukesneeringer** [comment]
CircleCI's explicit OS X support is a 1.0 thing.

**Comment by ofrobots** [comment]
What is the prioritization for this issue?

I have an interest in using a nightly build for catch regressions of the form in https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2238. This issue conflates testing on GCP and nightly testing. I don't have strong opinions of the former, but we need to fix the latter.

**Comment by lukesneeringer** [comment]
The prioritization is "this quarter", and I am driving the overall testing strategy for client libraries.
